File: bin/process_copc_conversions.php
<?php
declare(strict_types=1);

/**
 * Фоновая конвертация LAS → COPC (PR4, модуль 2.2). Без отдельной таблицы
 * очереди в БД — статус определяется наличием файлов на диске:
 *   foo.las
 *   foo.las.copc.laz          — результат (готов)
 *   foo.las.copc.lock         — конвертация уже запущена этим/предыдущим
 *                                запуском (см. bin/copc_convert_worker.ps1)
 *   foo.las.copc.error        — последняя попытка упала, не повторяем
 *                                автоматически (нужно разобраться вручную)
 *
 * Параллельность — не больше MAX_CONCURRENT процессов pdal.exe одновременно
 * (см. docs/CURRENT_STATE.md, раздел 10, п.4 — резерв под Apache/БД/NTRIP-
 * поллинг/RINEX-очередь на той же физической машине). Запускается по
 * расписанию (Windows Task Scheduler, как и bin/process_rinex_requests.php);
 * каждый запуск просто доливает работу до лимита и завершается — не ждёт
 * окончания уже запущенных конвертаций.
 */

require __DIR__ . '/../app/lib/db.php';
require __DIR__ . '/../app/lib/cli.php';
require_cli_or_token();

const MAX_CONCURRENT = 10;
// "Зависшая" конвертация — на диске обнаружен реальный случай: сервер
// перезагрузился/процесс убит на 53ГБ файле 13.07, .lock остался лежать
// БЕЗ .converting.copc.laz и без единой строки в логе — process_copc_
// conversions.php считал файл "уже конвертируется" и молча пропускал его
// КАЖДЫЙ запуск (каждые 2 минуты) больше двух недель подряд. Признак
// зависания — размер .converting.copc.laz не растёт дольше STALL_MINUTES
// (или файла нет вовсе спустя STARTUP_GRACE_MINUTES после создания lock —
// значит PDAL не успел запуститься/сразу упал). Прогресс между запусками
// сравнивается через .progress (см. ниже) — сам процесс PDAL его не
// пишет, это состояние ведёт только этот скрипт.
const STALL_MINUTES = 120;
const STARTUP_GRACE_MINUTES = 15;
// Сколько раз можно автоматически перезапустить один и тот же файл после
// обнаружения зависания, прежде чем сдаться и создать .error (как и для
// обычной ошибки PDAL — дальше нужно разбираться вручную, не долбить один
// и тот же файл бесконечно, если он стабильно виснет/падает).
const MAX_STALL_RETRIES = 3;

$uploadDir = realpath(__DIR__ . '/../uploads/tours') . DIRECTORY_SEPARATOR;
$psScript = __DIR__ . '/copc_convert_worker.ps1';
$logDir = __DIR__ . '/../uploads/copc_logs';
if (!is_dir($logDir)) {
    mkdir($logDir, 0777, true);
}

$pdo = db();
$lasFiles = [];
foreach ($pdo->query("SELECT file_path FROM tours WHERE file_format = 'las'") as $row) {
    $lasFiles[] = $row['file_path'];
}
foreach ($pdo->query("SELECT file_path FROM tour_files WHERE file_format = 'las'") as $row) {
    $lasFiles[] = $row['file_path'];
}
$lasFiles = array_unique($lasFiles);

// Проход 1: для КАЖДОГО файла с существующим .lock проверяем, не завис ли
// он — прежде чем считать его "уже запущенным" ниже. Делается на каждом
// запуске (каждые 2 минуты по расписанию), не только когда ищем новую
// работу — иначе зависший файл так и остался бы обнаружен только один раз.
foreach ($lasFiles as $relativePath) {
    $base = $uploadDir . $relativePath;
    $lockFile = $base . '.copc.laz.lock';
    if (!is_file($lockFile)) {
        continue;
    }
    $tmpOutput = $base . '.converting.copc.laz';
    $progressFile = $base . '.copc.laz.progress';
    $retriesFile = $base . '.copc.laz.stall_retries';

    $currentSize = is_file($tmpOutput) ? filesize($tmpOutput) : null;
    $prev = is_file($progressFile) ? json_decode((string)file_get_contents($progressFile), true) : null;

    $stalled = false;
    if ($currentSize === null) {
        // Ни одного байта выходного файла ещё нет — ждём STARTUP_GRACE_MINUTES
        // на случай, что PDAL просто ещё не успел начать писать (создание
        // lock и реальный старт процесса не строго одновременны).
        $lockAgeMin = (time() - filemtime($lockFile)) / 60;
        if ($lockAgeMin > STARTUP_GRACE_MINUTES) {
            $stalled = true;
            cli_err("Завис (нет выходного файла спустя " . round($lockAgeMin) . " мин): $relativePath");
        }
    } elseif ($prev !== null && $currentSize <= $prev['size'] && (time() - $prev['at']) / 60 > STALL_MINUTES) {
        // Файл есть, но не растёт уже дольше STALL_MINUTES с последней
        // проверки размера — процесс, скорее всего, мёртв (иначе PDAL
        // продолжал бы писать), но сам не снял свой lock (finally не
        // выполнился — kill/перезагрузка сервера).
        $stalled = true;
        cli_err("Завис (выходной файл не растёт дольше " . STALL_MINUTES . " мин): $relativePath");
    }

    if ($currentSize !== null && ($prev === null || $currentSize > $prev['size'])) {
        // Реальный прогресс с прошлой проверки — обновляем снимок.
        file_put_contents($progressFile, json_encode(['size' => $currentSize, 'at' => time()]));
    }

    if (!$stalled) {
        continue;
    }

    $retries = is_file($retriesFile) ? (int)file_get_contents($retriesFile) : 0;
    @unlink($lockFile);
    @unlink($tmpOutput);
    @unlink($progressFile);
    if ($retries + 1 >= MAX_STALL_RETRIES) {
        file_put_contents(
            $base . '.copc.laz.error',
            "Конвертация зависала $retries раз подряд (сервер перезапускался/процесс убит без завершения) — автоматические попытки остановлены, нужно разобраться вручную."
        );
        @unlink($retriesFile);
        cli_err("$relativePath: превышен лимит автоповторов ($retries), помечен как .error");
    } else {
        file_put_contents($retriesFile, (string)($retries + 1));
        cli_out("$relativePath: lock снят, будет перезапущен (попытка " . ($retries + 1) . "/" . MAX_STALL_RETRIES . ")");
    }
}

// Подсчитываем уже запущенные конвертации (lock-файл существует) по
// фактическим путям из БД — без glob по дереву каталогов (file_path может
// лежать в подпапке группы, см. комментарий в schema.sql про uploads/tours/
// <group-slug>/...), а не сканированием файловой системы.
$runningCount = 0;
foreach ($lasFiles as $relativePath) {
    if (is_file($uploadDir . $relativePath . '.copc.laz.lock')) {
        $runningCount++;
    }
}
cli_out("LAS-файлов в БД: " . count($lasFiles) . ", уже запущенных конвертаций (lock): $runningCount");

$spawned = 0;
foreach ($lasFiles as $relativePath) {
    if ($runningCount + $spawned >= MAX_CONCURRENT) {
        cli_out("Достигнут лимит параллельных конвертаций (" . MAX_CONCURRENT . ") — остальное в следующий запуск.");
        break;
    }

    $inputLas = $uploadDir . $relativePath;
    if (!is_file($inputLas)) {
        continue;
    }
    $outputCopc = $inputLas . '.copc.laz';
    $lockFile = $outputCopc . '.lock';
    $errorFile = $outputCopc . '.error';

    if (is_file($outputCopc) || is_file($lockFile) || is_file($errorFile)) {
        continue;
    }

    // Lock создаём ЗДЕСЬ (синхронно, до запуска процесса) — иначе при
    // нескольких файлах подряд в одном запуске мы могли бы посчитать их
    // все как "ещё не начатые" и превысить MAX_CONCURRENT до того, как
    // дочерние процессы успеют создать свои локи сами.
    touch($lockFile);

    $logFile = $logDir . '/' . basename($relativePath) . '.log';
    // Команда передаётся МАССИВОМ, а не строкой — со строкой proc_open на
    // Windows запускает через cmd.exe, и многослойное экранирование
    // (PHP escapeshellarg + cmd.exe + сам PowerShell) на путях с обратными
    // слэшами портило аргументы настолько, что скрипт падал ДО входа в
    // свой try/finally — lock-файл оставался висеть навечно. Массив минует
    // cmd.exe целиком (прямой CreateProcess), экранирование не нужно.
    $cmd = [
        'powershell.exe', '-NoProfile', '-ExecutionPolicy', 'Bypass', '-File', $psScript,
        '-InputLas', $inputLas,
        '-OutputCopc', $outputCopc,
        '-LockFile', $lockFile,
        '-ErrorFile', $errorFile,
    ];

    $descriptors = [
        0 => ['pipe', 'r'],
        1 => ['file', $logFile, 'a'],
        2 => ['file', $logFile, 'a'],
    ];
    $process = proc_open($cmd, $descriptors, $pipes);
    if ($process === false) {
        cli_err("Не удалось запустить конвертацию для $relativePath");
        @unlink($lockFile);
        continue;
    }
    // Не ждём (proc_close/блокирующее чтение pipes не вызываем) — процесс
    // продолжает работать после завершения этого скрипта; сам снимет lock
    // (см. copc_convert_worker.ps1, finally).
    fclose($pipes[0]);
    $spawned++;
    cli_out("Запущена конвертация: $relativePath -> {$relativePath}.copc.laz");
}

cli_out("Запущено новых конвертаций: $spawned");

File: pyserver/apps/tours/management/commands/process_copc_conversions.py
"""
Port of bin/process_copc_conversions.php. Background LAS -> COPC conversion
(no separate queue table -- status is determined by files on disk):
  foo.las
  foo.las.copc.laz          -- result (ready)
  foo.las.copc.laz.lock     -- conversion already running (this or a
                                previous run, see copc_convert_worker)
  foo.las.copc.laz.error    -- last attempt failed, not auto-retried
                                (needs manual investigation)

Concurrency -- no more than MAX_CONCURRENT pdal processes at once (room for
Postgres/NTRIP polling/RINEX queue on the same machine). Run on a schedule
(systemd timer); each run just tops up work to the limit and exits -- it
does not wait for already-running conversions to finish.
"""

import json
import subprocess
import sys
import time
from pathlib import Path

from django.conf import settings
from django.core.management.base import BaseCommand

from apps.tours.models import Tour, TourFile

MAX_CONCURRENT = 10
# A "stalled" conversion -- a real case seen on the old server: the server
# rebooted / the process was killed mid-conversion of a 53GB file, leaving
# .lock behind with NO .converting.copc.laz and not a single log line --
# every run (every 2 minutes) silently treated the file as "already
# converting" and skipped it for over two weeks straight. The signal for a
# stall is: the .converting.copc.laz size hasn't grown for longer than
# STALL_MINUTES (or the file doesn't exist at all STARTUP_GRACE_MINUTES
# after the lock was created -- meaning PDAL never got around to starting,
# or died immediately). Progress between runs is compared via a .progress
# sidecar (below) -- PDAL itself never writes it, only this command
# maintains that state.
STALL_MINUTES = 120
STARTUP_GRACE_MINUTES = 15
# How many times a stalled file can be auto-restarted before giving up and
# writing .error (same as for a normal PDAL error -- from then on it needs
# manual investigation, not endless hammering of a file that reliably
# stalls/fails).
MAX_STALL_RETRIES = 3


class Command(BaseCommand):
    help = "Spawn LAS -> COPC conversions up to MAX_CONCURRENT, with stall detection for lock files left behind by a killed/rebooted worker."

    def handle(self, *args, **options):
        upload_dir = Path(settings.UPLOADS_ROOT) / "tours"
        log_dir = Path(settings.UPLOADS_ROOT) / "copc_logs"
        log_dir.mkdir(parents=True, exist_ok=True)

        las_files = list(
            set(
                list(Tour.objects.filter(file_format="las").values_list("file_path", flat=True))
                + list(TourFile.objects.filter(file_format="las").values_list("file_path", flat=True))
            )
        )

        # Pass 1: for EVERY file with an existing .lock, check whether it's
        # stalled before treating it as "already running" below. Done on
        # every run (every couple of minutes on schedule), not only when
        # looking for new work -- otherwise a stalled file would only ever
        # be detected once.
        for relative_path in las_files:
            base = upload_dir / relative_path
            lock_file = Path(str(base) + ".copc.laz.lock")
            if not lock_file.is_file():
                continue
            tmp_output = Path(str(base) + ".converting.copc.laz")
            progress_file = Path(str(base) + ".copc.laz.progress")
            retries_file = Path(str(base) + ".copc.laz.stall_retries")

            current_size = tmp_output.stat().st_size if tmp_output.is_file() else None
            prev = None
            if progress_file.is_file():
                try:
                    prev = json.loads(progress_file.read_text())
                except (OSError, ValueError):
                    prev = None

            stalled = False
            if current_size is None:
                # Not a single byte of output yet -- wait STARTUP_GRACE_MINUTES
                # in case PDAL just hasn't started writing yet (the lock being
                # created and the process actually starting aren't perfectly
                # simultaneous).
                lock_age_min = (time.time() - lock_file.stat().st_mtime) / 60
                if lock_age_min > STARTUP_GRACE_MINUTES:
                    stalled = True
                    self.stderr.write(f"Р—Р°РІРёСЃ (РЅРµС‚ РІС‹С…РѕРґРЅРѕРіРѕ С„Р°Р№Р»Р° СЃРїСѓСЃС‚СЏ {round(lock_age_min)} РјРёРЅ): {relative_path}")
            elif prev is not None and current_size <= prev["size"] and (time.time() - prev["at"]) / 60 > STALL_MINUTES:
                # Output file exists but hasn't grown in over STALL_MINUTES
                # since the last check -- the process is most likely dead
                # (otherwise PDAL would keep writing) but never removed its
                # own lock (its cleanup path never ran -- kill/reboot).
                stalled = True
                self.stderr.write(f"Р—Р°РІРёСЃ (РІС‹С…РѕРґРЅРѕР№ С„Р°Р№Р» РЅРµ СЂР°СЃС‚С‘С‚ РґРѕР»СЊС€Рµ {STALL_MINUTES} РјРёРЅ): {relative_path}")

            if current_size is not None and (prev is None or current_size > prev["size"]):
                # Real progress since the last check -- update the snapshot.
                progress_file.write_text(json.dumps({"size": current_size, "at": time.time()}))

            if not stalled:
                continue

            retries = int(retries_file.read_text()) if retries_file.is_file() else 0
            lock_file.unlink(missing_ok=True)
            tmp_output.unlink(missing_ok=True)
            progress_file.unlink(missing_ok=True)
            if retries + 1 >= MAX_STALL_RETRIES:
                Path(str(base) + ".copc.laz.error").write_text(
                    f"РљРѕРЅРІРµСЂС‚Р°С†РёСЏ Р·Р°РІРёСЃР°Р»Р° {retries} СЂР°Р· РїРѕРґСЂСЏРґ (СЃРµСЂРІРµСЂ РїРµСЂРµР·Р°РїСѓСЃРєР°Р»СЃСЏ/РїСЂРѕС†РµСЃСЃ СѓР±РёС‚ Р±РµР· Р·Р°РІРµСЂС€РµРЅРёСЏ) "
                    "вЂ” Р°РІС‚РѕРјР°С‚РёС‡РµСЃРєРёРµ РїРѕРїС‹С‚РєРё РѕСЃС‚Р°РЅРѕРІР»РµРЅС‹, РЅСѓР¶РЅРѕ СЂР°Р·РѕР±СЂР°С‚СЊСЃСЏ РІСЂСѓС‡РЅСѓСЋ.",
                    encoding="utf-8",
                )
                retries_file.unlink(missing_ok=True)
                self.stderr.write(f"{relative_path}: РїСЂРµРІС‹С€РµРЅ Р»РёРјРёС‚ Р°РІС‚РѕРїРѕРІС‚РѕСЂРѕРІ ({retries}), РїРѕРјРµС‡РµРЅ РєР°Рє .error")
            else:
                retries_file.write_text(str(retries + 1))
                self.stdout.write(f"{relative_path}: lock СЃРЅСЏС‚, Р±СѓРґРµС‚ РїРµСЂРµР·Р°РїСѓС‰РµРЅ (РїРѕРїС‹С‚РєР° {retries + 1}/{MAX_STALL_RETRIES})")

        # Count already-running conversions (lock file exists) by the actual
        # DB paths -- not by globbing the directory tree (file_path can live
        # in a group subfolder, see the schema.sql comment on
        # uploads/tours/<group-slug>/...).
        running_count = sum(
            1 for relative_path in las_files
            if Path(str(upload_dir / relative_path) + ".copc.laz.lock").is_file()
        )
        self.stdout.write(f"LAS-С„Р°Р№Р»РѕРІ РІ Р‘Р”: {len(las_files)}, СѓР¶Рµ Р·Р°РїСѓС‰РµРЅРЅС‹С… РєРѕРЅРІРµСЂС‚Р°С†РёР№ (lock): {running_count}")

        spawned = 0
        for relative_path in las_files:
            if running_count + spawned >= MAX_CONCURRENT:
                self.stdout.write(f"Р”РѕСЃС‚РёРіРЅСѓС‚ Р»РёРјРёС‚ РїР°СЂР°Р»Р»РµР»СЊРЅС‹С… РєРѕРЅРІРµСЂС‚Р°С†РёР№ ({MAX_CONCURRENT}) вЂ” РѕСЃС‚Р°Р»СЊРЅРѕРµ РІ СЃР»РµРґСѓСЋС‰РёР№ Р·Р°РїСѓСЃРє.")
                break

            input_las = upload_dir / relative_path
            if not input_las.is_file():
                continue
            output_copc = str(input_las) + ".copc.laz"
            lock_file = Path(output_copc + ".lock")
            error_file = Path(output_copc + ".error")

            if Path(output_copc).is_file() or lock_file.is_file() or error_file.is_file():
                continue

            # Create the lock HERE (synchronously, before spawning the
            # process) -- otherwise, with several files in one run, they
            # could all be counted as "not started yet" and MAX_CONCURRENT
            # could be exceeded before the child processes get a chance to
            # create their own locks.
            lock_file.touch()

            log_file = log_dir / f"{Path(relative_path).name}.log"
            with open(log_file, "a") as log_fh:
                subprocess.Popen(
                    [
                        sys.executable, str(Path(settings.BASE_DIR) / "manage.py"), "copc_convert_worker",
                        "--input-las", str(input_las),
                        "--output-copc", output_copc,
                        "--lock-file", str(lock_file),
                        "--error-file", str(error_file),
                    ],
                    stdin=subprocess.DEVNULL, stdout=log_fh, stderr=log_fh,
                    start_new_session=True,
                )
            spawned += 1
            self.stdout.write(f"Р—Р°РїСѓС‰РµРЅР° РєРѕРЅРІРµСЂС‚Р°С†РёСЏ: {relative_path} -> {relative_path}.copc.laz")

        self.stdout.write(f"Р—Р°РїСѓС‰РµРЅРѕ РЅРѕРІС‹С… РєРѕРЅРІРµСЂС‚Р°С†РёР№: {spawned}")

